File: backend/app/models/meal_delivery.py
import uuid
from datetime import date, datetime

from sqlalchemy import Date, DateTime, Enum, ForeignKey, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base
from app.models._mixins import GUID, TimestampMixin, uuid_pk
from app.models.meal_skip import MealType


class MealDelivery(Base, TimestampMixin):
    __tablename__ = "meal_deliveries"
    __table_args__ = (
        UniqueConstraint("user_id", "date", "meal_type", name="uq_meal_delivery_user_date_meal"),
    )

    id: Mapped[uuid.UUID] = uuid_pk()
    user_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    meal_type: Mapped[MealType] = mapped_column(
        Enum(MealType, name="meal_type"), nullable=False
    )
    delivered_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    delivered_by: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("users.id"), nullable=False
    )

File: frontend/lib/core/models/meal_delivery.dart
class MealDelivery {
  final String id;
  final String userId;
  final DateTime date;
  final String mealType;
  final DateTime deliveredAt;
  final String deliveredBy;

  MealDelivery({
    required this.id,
    required this.userId,
    required this.date,
    required this.mealType,
    required this.deliveredAt,
    required this.deliveredBy,
  });

  factory MealDelivery.fromJson(Map<String, dynamic> json) => MealDelivery(
        id: json['id'] as String,
        userId: json['user_id'] as String,
        date: DateTime.parse(json['date'] as String),
        mealType: json['meal_type'] as String,
        deliveredAt: DateTime.parse(json['delivered_at'] as String),
        deliveredBy: json['delivered_by'] as String,
      );
}
